pkg/controller/awsfederatedrole/awsfederatedrole_controller.go

```diff
@@ -0,0 +1,269 @@
+
+/**
+* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
+* business logic.  Delete these comments after modifying this file.*
+ */
+
+// Add creates a new AWSFederatedRole Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileAWSFederatedRole{client: mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		awsClientBuilder: awsclient.NewClient}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("awsfederatedrole-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource AWSFederatedRole
+	err = c.Watch(&source.Kind{Type: &awsv1alpha1.AWSFederatedRole{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileAWSFederatedRole implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileAWSFederatedRole{}
+
+// ReconcileAWSFederatedRole reconciles a AWSFederatedRole object
+type ReconcileAWSFederatedRole struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client           client.Client
+	scheme           *runtime.Scheme
+	awsClientBuilder func(awsAccessID, awsAccessSecret, token, region string) (awsclient.Client, error)
+}
+
+// Reconcile reads that state of the cluster for a AWSFederatedRole object and makes changes based on the state read
+// and what is in the AWSFederatedRole.Spec
+func (r *ReconcileAWSFederatedRole) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling AWSFederatedRole")
+
+	// Fetch the AWSFederatedRole instance
+	instance := &awsv1alpha1.AWSFederatedRole{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If the CR is known to be Valid or Invalid, doesn't need to be reconciled.
+	if instance.Status.State == awsv1alpha1.AWSFederatedRoleStateValid || instance.Status.State == awsv1alpha1.AWSFederatedRoleStateInvalid {
```

Comment:
What happens if an instance is updated ? Would this stop them from being processed or is this not part of the "MVP" ?

Reply:
Yes this would not allow it to be processed unless the status is removed. This is intentional. If the status is invalid or valid, there is no way to see what changed if the CR is updated. And technically, the status of the updated CR is no longer valid or invalid.

Reply:
This also prevents an invalid policy from constantly going through the reconcile.